libteec: implement OCALLs

HernanGatta opened the pull request:
# Overview
This PR implements OCALLs in the TEE Client API.

**Related PRs:** [OP-TEE](https://github.com/OP-TEE/optee_os/pull/3673), [Linux](https://github.com/linaro-swg/linux/pull/72), [OP-TEE Examples](https://github.com/linaro-swg/optee_examples/pull/64).

## Session Open
A new function is added, `TEEC_OpenSessionEx`, whose signature is identical to `TEEC_OpenSession`, but takes two additional parameters. The first is an array of typed settings, while the second provides the number of settings in the array.

One such setting is the OCALL setting, which allows the developer to specify a callback for whenever an OCALL arrives, as well as an arbitrary context pointer.

The rationale behind using session settings is that it is conceivable that in the future, additional features are added that require configuration. Instead of creating new, independent functions to deal with those new features, a new setting type can be added and the existing API set remains the same. The Open Enclave SDK implements a similar pattern for its [`oe_create_enclave`](https://github.com/openenclave/openenclave/blob/master/include/openenclave/host.h#L149) function.

When the OCALL setting is specified in `TEEC_OpenSessionEx`, it checks that the underlying context reports that OP-TEE supports OCALLs. If so, it opens a session in the usual way via `TEEC_OpenSession`, then stores the OCALL configuration in the session object.

Sample usage:

```
TEEC_SessionSettingOcall ocall_setting = { my_ocall_handler, &my_context };
TEEC_SessionSetting settings[] = {
        { .type = TEEC_SESSION_SETTING_OCALL, .u.ocall = &ocall_setting }
};

res = TEEC_OpenSessionEx(&ctx, &sess, &uuid, TEEC_LOGIN_PUBLIC, NULL,
                         NULL, &err_origin, settings, 1);
```

## OCALL Handling
When OCALL support is configured, calls to `TEEC_InvokeFunction` are rerouted via the ECALL IOCTL instead of the INVOKE IOCTL. Around the IOCTL invocation, a loop exists to catch early returns, indicating that an OCALL request arrived. When that happens, the Client API takes care of dealing with shared memory allocation and free requests, as well as of parameter processing before and after dispatching an OCALL to the OCALL handler.

Signed-off-by: <hegatta@microsoft.com>

HernanGatta commented:
@etienne-lms, thank you for the review.

Regarding the renaming of `TEEC_OpenSessionEx` to `TEEC_OpenSessionWithRpc`, I would argue that there is nothing in the signature of the new function that is RPC/OCALL-specific. The settings mechanism is generic and could be used for other session settings in the future, so I'd be reluctant to rename the function to something that isn't also as generic. 

The `Ex` suffix comes from my time working on Windows, where C APIs that are overloaded over time pick up the `Ex`tended suffix. I wasn't sure what else to call the new function, so I tacked `Ex` at the end; I'm all for renaming the function to something else.

Perhaps `TEEC_OpenSessionWithSettings`?

With respect to replacing "OCALL" with "RPC", I'm not opposed. However, @jenswi-linaro had expressed a liking for the name "OCALL" in [the issue](https://github.com/OP-TEE/optee_os/issues/3303#issuecomment-543212784) we discussed this feature in last year. Additionally, OCALL, or "out call", is the term used by the Open Enclave SDK and Intel's SGX for the same feature, so I figured we could reuse it.

Having said that, I'll go with whatever name you all coalesce on that fits within the readily established nomenclature.

Lastly, as to your request for a cleaner patch set, I'll gladly readjust the commits to fit what you requested, though I'll wait until we've decided on the right names.

etienne-lms commented:
`TEEC_OpenSessionWithSettings` is fine also.
As for OCalls/ECalls, keep these naming if they reference to some standard. Just make sure each is described in some header file.
As for the capability, `TEE_OPTEE_CAP_CA_OCALL` to route OCall to CA provided handlers, no?

Regarding, the `TEEC_OpenSessionExtended()`, @jenswi-linaro mentioned registering OCall handlers from the context. Do you think each session should provide the OCall handler? Or one could register OCall settings to the `TEEC_Context` created, applicable to all sessions created in the context.

HernanGatta commented:
> As for the capability, `TEE_OPTEE_CAP_CA_OCALL` to route OCall to CA provided handlers, no?
Yes.

I'm of two minds about adding this capability. On the one hand, it's possible that the version of OP-TEE that the client library is talking to does not support OCALLs. In this case, we fail right away if the CA expects OCALL support but OP-TEE does not provide it. On the other hand, we could do away with the capability, and just open the session. If the CA expects OCALL support, this means that the TA probably makes use of OCALLs. If OP-TEE was compiled without OCALL support, the TA will fail to make OCALLs, and that failure will presumably be reported back to the CA.

So we have the choice between checking for OCALL support and failing right away before even opening the session, or not checking for OCALL support and letting the TA fail to send any later.

I chose to be explicit and have OP-TEE report support, or lack thereof, for the feature. However, if we wish to not pile on another capability, it's possible to do without it.

What do you think?

> Regarding, the `TEEC_OpenSessionExtended()`, @jenswi-linaro mentioned registering OCall handlers from the context. Do you think each session should provide the OCall handler? Or one could register OCall settings to the `TEEC_Context` created, applicable to all sessions created in the context.

I've considered this, and I'm open to changing my mind.

The way I conceptualize OCALLs is as a response to a function invocation. That is, one calls `TEEC_InvokeFunction` and one might get an OCALL in return. Since the former is called on a session and not a context, it seemed logical to me to attach OCALLs to sessions. An OCALL is a response to a function invocation in the context of an established session.

Additionally, a developer presumably controls both the CA and the TA. As such, when opening a session to a TA that is known to request OCALLs, it would make sense that the conduit for delivery is the session; OCALLs at the context level may be conceptualized as coming from any TA to which a session is opened via that context, regardless of whether all or only some of the TAs request OCALLs.

Lastly, OCALLs must still be attached to sessions, at least technically. This is because on session close, all outstanding OCALLs must be cancelled. If one prematurely closes a session, say due to an error, we can't wait until the context is released to cancel pending OCALLs because the secure thread is stuck on an RPC.

The flip-side to these arguments is an observation that Jens made whereby a TA may call another TA which itself sends an OCALL. Therefore, the CA might receive an OCALL from a TA other than the one the session was opened against. This is the reason I added the TA UUID parameter to the OCALL handler callback, which wasn't there in the first try at this feature. In light of this, it could make sense to have a catch-all OCALL handler in the CA for all possible TAs. Hence, if the handler must already support handling OCALLs from multiple TAs, it may as well handle OCALLs from all TAs opened against a context, at which point the handler may also as well be attached to the context.

The counterargument the previous argument is that a developer might still want to know which session an OCALL came from and decide what to do accordingly, especially if multiple sessions are opened on the same context against the same multi-instance TA: the handler will probably want to know which session the OCALL came from to distinguish between TA instances. This could be resolved by adding a session parameter to the OCALL handler callback (which I think I should add anyway).

Thoughts? I'm currently leaning toward the second line of thought and moving the callback handler setting to the context, as long as the callback handler has a parameter that carries the session from which the OCALL originated.

Lastly, Jens also mentioned that one is not allowed to call back into the TA from the OCALL handler, which is true, and I've added no way to detect this. I'll go ahead and implement a check to prevent re-entry.

jforissier commented:
Adding my 2 (euro) cents to the discussion...

- Regarding the handler: considering all the arguments above, I have a slight preference for attaching the callback to the context rather than the session.
- I'm fine with name `TEEC_OpenSessionEx()` (brings back some memories! :wink:) although I have no strong opinion. TBH `WithSettings` does not convey much more information than `Ex`. GP add a `2` when they evolve their APIs (I'm thinking Internal Core), so at least there would be no conflicts in both cases.
- I am also in favor of having a capability flag and failing early if OP-TEE does not support the feature.

jenswi-linaro commented:
> So we have the choice between checking for OCALL support and failing right away before even opening the session, or not checking for OCALL support and letting the TA fail to send any later.
> 
> I chose to be explicit and have OP-TEE report support, or lack thereof, for the feature. However, if we wish to not pile on another capability, it's possible to do without it.

I prefer failing early, it's a bit more tidy in my mind.

Registering OCALL handler in the context or the session?
I'm leaning towards context since I consider a global thing. It will have to be able to serve all TAs in a family of TAs regardless of to which TA the session in opened since a TA may call another TA. The OCALL handler will need to know which session it's serving. It would be handy to have a session specific pointer too. We could add TEEC_SetSessionSpecificPointer() and TEEC_GetSessionSpecificPointer(). I think we should avoid the name context here since we already have a TEEC_Context.

HernanGatta commented:
I have some changes incoming to this one, with simplifications, too.

**Design question:**

Currently, the PR has the TEE Client API pass shared memory to the CA that was allocated on request of the TA via the `memref` element of `TEEC_Parameter`.

While this reflects the underlying fact that the shared memory in question can be used across multiple OCALLs (i.e., it's not a `tmpref` whose lifetime is tied to the current OCALL), it does put the burden of dealing with offsets into the parent shared memory buffer on the OCALL handler, and thus on the CA developer. It also exposes an implementation detail, namely that the current OCALL implementation puts all buffers passed from the TA to the CA into the same shared memory element. As a result, what starts out in the TA as a buffer-size pair into `TEE_InvokeCACommand` winds up on the CA side as a buffer-size-offset triplet.

This is not how function invocations on TAs work: regardless of whether the underlying shared memory can persist across multiple invocations or is temporary, the TA function invocation handler gets a buffer and a size; no offsets.

The changes I'm working on would instead have the TEE Client API pass buffers to the CA's OCALL handler as `tmpref`s, removing the burden of dealing with the offset from the OCALL handler.

I'm torn: on the one hand, using temporary reference semantics makes it easier to write OCALL handlers because offsets are not involved, and don't have to be, since `libteec` can perform the pointer arithmetic on behalf of the developer. On the other, it's semantically incorrect.

What do you all think?

HernanGatta commented on libteec/include/linux/tee.h:
```diff
@@ -63,6 +64,7 @@
  * OP-TEE specific capabilities
  */
 #define TEE_OPTEE_CAP_TZ	(1 << 0)
+#define TEE_OPTEE_CAP_OCALL	(1 << 1)/* Supports calls from TA to CA */
```
I've turned it into a generic one, and is now handled as such in the library.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -562,6 +974,58 @@ TEEC_Result TEEC_OpenSession(TEEC_Context *ctx, TEEC_Session *session,
+		}
+	}
+
+	res = TEEC_OpenSession(ctx, session, destination, connection_method,
```
This has been addressed by attaching the OCALL handler to the context, as you've requested.

HernanGatta commented on public/tee_client_api.h:
```diff
@@ -246,6 +249,14 @@ extern "C" {
+ * @param p The paramType.
+ * @param i The i-th parameter to set the type for.
+ */
+#define TEEC_PARAM_TYPE_SET(t, i) (((uint32_t)(t) & 0xF) << ((i)*4))
```
Mm... I'd have to steal `GENMASK` from the kernel, unless it's in some header I'm not familiar with. I don't know if that'd be clearer.

Perhaps I can take `0xF` out and give it a name that describes why that number?

HernanGatta commented:
> I prefer failing early, it's a bit more tidy in my mind.

Checked during context initialization.
 
> Registering OCALL handler in the context or the session?
> I'm leaning towards context since I consider this a global thing. It will have to be able to serve all TAs in a family of TAs regardless of to which TA the session in opened since a TA may call another TA. The OCALL handler will need to know which session it's serving. It would be handy to have a session specific pointer too. We could add TEEC_SetSessionSpecificPointer() and TEEC_GetSessionSpecificPointer(). I think we should avoid the name context here since we already have a TEEC_Context.

I've added the handler and a context-specific data pointer as a setting on `TEEC_Context` as well as a session-specific pointer also as a setting on `TEEC_Session`.

jforissier commented:
> P.S.: The Travis CI failure is about style complaints that I don't think are true issues.

I think there is only one warning type that we want to ignore here ("do not add new typedefs"), but all the others can be eliminated:
- "Missing blank line" is caused by `SLIST_ENTRY(teec_shm_record)` not being recognized as a type. Can be fixed by adding `SLIST_ENTRY\(.*\)` to `typedefs.checkpatch`
- "Alignment should match open parenthesis": align as expected
- "spaces preferred around that '*'": add the new types to `typedefs.checkpatch`: `TEEC_ContextSetting`, `TEEC_ContextSettingOcall`, `TEEC_SessionSetting` and `TEEC_SessionSettingData`.

HernanGatta commented:
> * "Missing blank line" is caused by `SLIST_ENTRY(teec_shm_record)` not being recognized as a type. Can be fixed by adding `SLIST_ENTRY\(.*\)` to `typedefs.checkpatch`
> * "spaces preferred around that '*'": add the new types to `typedefs.checkpatch`: `TEEC_ContextSetting`, `TEEC_ContextSettingOcall`, `TEEC_SessionSetting` and `TEEC_SessionSettingData`.

Ah, thanks for the pointers!

> * "Alignment should match open parenthesis": align as expected

I'll happily do so, but the style won't match the rest of the file if I do. Parameters are indented with three tabs in the file, for example:

```
static void teec_post_process_operation(TEEC_Operation *operation,
			struct tee_ioctl_param *params,
			TEEC_SharedMemory *shms)
...
```

Preference?

HernanGatta commented on libteec/include/linux/tee.h:
```diff
@@ -243,6 +250,33 @@ struct tee_ioctl_open_session_arg {
+ * Command sent to the CA to request allocation of shared memory to carry the
+ * parameters of an OCALL
+ *
+ * [out] param[0].u.value.a		SHM ID
```
Wrong.

HernanGatta commented on libteec/include/linux/tee.h:
```diff
@@ -243,6 +250,33 @@ struct tee_ioctl_open_session_arg {
+/*
+ * Command sent to the CA to free previously allocated shared memory.
+ *
+ * [in] param[0].u.value.a		SHM ID
```
Idem, will fix.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -51,12 +52,33 @@
+#define SHIFT_U32(v, shift)	((uint32_t)(v) << (shift))
+#define PTR_ADD(p1, p2)		((void *)((uintptr_t)(p1) + (uintptr_t)(p2)))
+
+/* This only exists to avoid including <sys/queue.h> in tee_client_api.h */
+struct teec_shm_record {
+	TEEC_SharedMemory shm;
+	SLIST_ENTRY(teec_shm_record) next;
+};
```
I didn't know if you'd regard including of `<sys/queue.h>` in `tee_client_api.h` as a potential problem, so I did this. Happy to change it.

jforissier replied:
Fine with me. It is preferable not to pull too many dependencies in the client header file I think.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -573,6 +660,55 @@ TEEC_Result TEEC_OpenSession(TEEC_Context *ctx, TEEC_Session *session,
 	return res;
 }
 
+TEEC_Result TEEC_OpenSession2(TEEC_Context *context,
+			      TEEC_Session *session,
+			      const TEEC_UUID *destination,
+			      uint32_t connectionMethod,
+			      const void *connectionData,
+			      TEEC_Operation *operation,
+			      uint32_t *returnOrigin,
+			      const TEEC_SessionSetting *settings,
+			      uint32_t numSettings)
```
Re-indent to three tabs?

jforissier commented:
> > * "Alignment should match open parenthesis": align as expected
> 
> I'll happily do so, but the style won't match the rest of the file if I do. Parameters are indented with three tabs in the file, for example:
> 
> ```
> static void teec_post_process_operation(TEEC_Operation *operation,
> 			struct tee_ioctl_param *params,
> 			TEEC_SharedMemory *shms)
> {
> 	...
> ```
> 
> Preference?

As much as I like consistency :wink: I'd rather write new code "the proper way" in this case.

HernanGatta commented:
> As much as I like consistency  I'd rather write new code "the proper way" in this case.

All lined up now.

> I think there is only one warning type that we want to ignore here ("do not add new typedefs"), but all the others can be eliminated

We now only have that one, and one `CHECK` about the `BIT` macro, which was briefly discussed in the [driver PR](https://github.com/linaro-swg/linux/pull/72#pullrequestreview-394449753). The issue is that that macro isn't exported to user-space, so I went back to shifts.

HernanGatta commented on public/tee_client_api.h:
```diff
@@ -198,11 +198,14 @@ extern "C" {
 #define TEEC_ORIGIN_COMMS        0x00000002
 #define TEEC_ORIGIN_TEE          0x00000003
 #define TEEC_ORIGIN_TRUSTED_APP  0x00000004
+#define TEEC_ORIGIN_CLIENT_APP   0x00000005
```
Mm, that's an excellent point, I didn't think of that. I like your suggestion; I'll update the PR accordingly (assuming no-one disagrees).

HernanGatta replied:
Done.

HernanGatta commented on public/tee_client_api.h:
```diff
@@ -374,6 +374,104 @@ typedef union {
+	bool memref_null;
+	bool ocall;
+	TEEC_ContextSettingOcall ocall_setting;
+} TEEC_Context;
```
Another good point. We can always store the settings in a separate structure and match them by pointer (i.e., in a dictionary). It wouldn't be terribly pretty, but would solve this issue if we'd rather not rev. the version.

Thanks for checking the spec.

jenswi-linaro replied:
Makes sense to increase the version number.

HernanGatta replied:
Done

etienne-lms commented:
This P-R needs to be rebased.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -163,13 +163,65 @@ TEEC_Result TEEC_InitializeContext(const char *name, TEEC_Context *ctx)
+		return res;
+
+	if (!settings)
+		return res;
```
Done.

HernanGatta commented on public/tee_client_api.h:
```diff
@@ -198,11 +198,14 @@ extern "C" {
  * TEEC_ORIGIN_TRUSTED_APP  The error originated within the Trusted Application
  *                          code.
+ * TEEC_ORIGIN_CLIENT_APP   The error originated within the Client Application
+ *                          code.
```
Done.

HernanGatta commented:
> This P-R needs to be rebased.

Done.

---

On a separate note, I noticed warnings still about typedefs from `checkpatch` in Travis CI. It's curious because each commit that adds a new typedef also adds the typedefs in question to `typedefs.checkpatch`.

Any ideas?

jforissier commented:
> On a separate note, I noticed warnings still about typedefs from `checkpatch` in Travis CI. It's curious because each commit that adds a new typedef also adds the typedefs in question to `typedefs.checkpatch`.
> 
> Any ideas?

The purpose of `typedefs.checkpatch` is not to prevent this particular warning but to avoid later errors that would occur if the new type was not known to checkpatch. In other words, it allows correct parsing of the code later on. So this is good: the tool raised our attention, we chose to ignore the warning, and we don't expect any new warning when using the new types.

jforissier commented:
One minor thing re. commit "checkpatch: reverse sort typedefs.checkpatch", and although I have reviewed it already... I suggest the following additional change:

```diff
diff --git a/typedefs.checkpatch b/typedefs.checkpatch
index 9b7906d..729ae36 100644
--- a/typedefs.checkpatch
+++ b/typedefs.checkpatch
@@ -69,9 +69,9 @@ CK_FLAGS
 CK_DESTROYMUTEX
 CK_DATE_PTR
 CK_DATE
-CK_CREATEMUTEX
 CK_C_INITIALIZE_ARGS_PTR
 CK_C_INITIALIZE_ARGS
+CK_CREATEMUTEX
 CK_CHAR_PTR
 CK_CHAR
 CK_CCM_PARAMS_PTR
```

Indeed, `sort -r` (as recommended in the comment) depends on the current locale and I'd rather we use the `C` locale, which follows traditional byte order. For example, consider this:
```
$ printf "CK_CREATEMUTEX\nCK_C_INITIALIZE_ARGS_PTR\nCK_CHAR_PTR" | LC_ALL=C sort -r
CK_C_INITIALIZE_ARGS_PTR
CK_CREATEMUTEX
CK_CHAR_PTR
$ printf "CK_CREATEMUTEX\nCK_C_INITIALIZE_ARGS_PTR\nCK_CHAR_PTR" | LC_ALL=en_US.UTF-8 sort -r
CK_CREATEMUTEX
CK_C_INITIALIZE_ARGS_PTR
CK_CHAR_PTR
```
Note, this "C" ordering happens to be what I obtained in `vim` using the `:sort!` command applied to all lines but the inital comments (although, according to `:help sort`, there is no guarantee).

In any case there does not seem to be any impact on checkpatch -- the important thing is to have `foo<something>` come before `foo` (in reverse sort) [1] which I guess is true for all locales (?).

[1] As I said https://github.com/OP-TEE/optee_os/pull/2685#issue-235520026 and this is still true with `checkpatch.pl` from kernel v5.5

jforissier commented on libteec/CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 project(libteec C)
 
-set(PROJECT_VERSION "1.0.0")
+set(PROJECT_VERSION "2.0.0")
```
`SOVERSION` needs to be set to `2` below

HernanGatta replied:
Good catch! I did the same thing as in the `Makefile` with the major, minor, and patch variables to avoid this problem going forward.

HernanGatta commented:
> I suggest the following additional change:

Done.

jforissier commented on public/tee_client_api.h:
```diff
@@ -374,6 +363,72 @@ typedef union {
 } TEEC_Parameter;
 
+/**
+ * TEEC_Result (*TEEC_OcallHandler) - Type for a CA-provided function to call
```
Minor: wouldn't CamelCase for OCALL be `OCall` rather than `Ocall`? (the O and the C are initials for individual words).

HernanGatta replied:
Indeed. Fixed!

HernanGatta commented on public/tee_client_api.h:
```diff
@@ -374,6 +374,104 @@ typedef union {
+ * OCALLs.
+ *
+ * @param handler  Pointer to the function to execute to handle an OCALL.
+ * @param data     Arbitrary pointer to pass to the OCALL handler function.
```
Yes. I've added comments indicating this.

HernanGatta commented on public/tee_client_api_extensions.h:
```diff
@@ -50,6 +50,64 @@ TEEC_Result TEEC_RegisterSharedMemoryFileDescriptor(TEEC_Context *context,
+ *                     to. If name is set to NULL, the default TEE is connected
+ *                     to. NULL is the only supported value in this version of
+ *                     the API implementation.
+ * @param context      The context structure which is to be initialized.
```
I actually made a mistake in the header declaration. Everywhere else the parameter is named `context`, so I've changed `ctx` to that.

HernanGatta commented on public/tee_client_api_extensions.h:
```diff
@@ -50,6 +50,64 @@ TEEC_Result TEEC_RegisterSharedMemoryFileDescriptor(TEEC_Context *context,
 
+/**
+ * TEEC_InitializeContext2() - Initializes a context holding connection
+ * information on the specific TEE, designated by the name string.
```
Edited as requested.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -157,13 +183,63 @@ TEEC_Result TEEC_InitializeContext(const char *name, TEEC_Context *ctx)
+	uint32_t n;
+	TEEC_Result res;
+
+	if ((!settings && numSettings) || (settings && !numSettings))
```
Fixed.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -157,13 +183,63 @@ TEEC_Result TEEC_InitializeContext(const char *name, TEEC_Context *ctx)
+	if ((!settings && numSettings) || (settings && !numSettings))
+		return TEEC_ERROR_BAD_PARAMETERS;
+
+	if (settings) {
```
Removed.

HernanGatta commented on libteec/src/tee_client_api.c:
```diff
@@ -51,12 +52,35 @@
 #define TEEC_MAX_DEV_SEQ	10
 
+/* No SHM flags */
+#define TEE_IOCTL_SHM_NONE	0
```
Added it to the driver PR.

HernanGatta commented:
> In the description for "libteec: implement OCALL support during function invocation":
> 
> * s/TEEC_InvokeFunction/TEEC_InvokeCommand()/
> * It could be worth mentioning explicitly that the OCALL handler is executed in the context of the thread that called `TEEC_InvokeCommand()` (did I get this right? I think I did.)
> 
> Plus one minor thing below. With that:
> 
> `Reviewed-by: Jerome Forissier <jerome@forissier.org>`

Thank you @jforissier and @etienne-lms  for your reviews, and my sincere apologies for the massive delay. I believe I've addressed your comments. I've also modified the description of the commit as Jérôme requested.

jforissier commented:
> 
> Thank you @jforissier and @etienne-lms for your reviews, and my sincere apologies for the massive delay.

Hi @HernanGatta, it's good to hear from you again! :+1: :smile:

To be honest, I was afraid that you would have moved on to something else, which would have been unfortunate given the amount of work already done on this feature (especially on your side!). That being said, don't worry about the delay. We are not on a tight schedule, and contributions and reviews are always made on a best effort basis. It should be pretty clear to everyone, hopefully.

> I believe I've addressed your comments. As to Jérôme's comments on the commit description, I've edited it as requested (and you are indeed correct on the thread context comment.)
> 
> I've foregone adding the `Reviewed-by` tag since it's been a while and you might have more comments.

Thanks. I've double checked, it all looks good and well documented, so:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

HernanGatta commented:
> To be honest, I was afraid that you would have moved on to something else, which would have been unfortunate given the amount of work already done on this feature (especially on your side!). That being said, don't worry about the delay. We are not on a tight schedule, and contributions and reviews are always made on a best effort basis. It should be pretty clear to everyone, hopefully.

Thank you for your note. There's been a bit of a switcheroo regarding work and priorities but I'm not letting this go :) 

> Thanks. I've double checked, it all looks good and well documented, so:
> 
> `Reviewed-by: Jerome Forissier <jerome@forissier.org>`

Applied, thanks!

HernanGatta commented:
Latest push is a rebase only.
